### PR TITLE
Resubmit #63279

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 53af3d6ea63d2710fa33ce592f5ba092ec3054ac with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69861/53af3d6ea63d2710fa33ce592f5ba092ec3054ac/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/69861/53af3d6ea63d2710fa33ce592f5ba092ec3054ac/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by nikitamikhaylov** [comment]
@vdimir It also looks like this PR broke some tests while was in master https://s3.amazonaws.com/clickhouse-test-reports/0/1b77b438fdfa44f9641c6a65fc4dcf5e6ad82fb5/integration_tests__asan__[3_4]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel1_0.log

Let's make sure it won't happen again after merge...

File: src/Core/SettingsChangesHistory.cpp [hunk 1]
```diff
@@ -85,7 +85,8 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
...
             {"join_to_sort_minimum_perkey_rows", 0, 40, "The lower limit of per-key average rows in the right table to determine whether to rerange the right table by key in left or inner join. This setting ensures that the optimization is not applied for sparse table keys"},
             {"join_to_sort_maximum_table_rows", 0, 10000, "The maximum number of rows in the right table to determine whether to rerange the right table by key in left or inner join"},
-            {"allow_experimental_join_right_table_sorting", false, false, "If it is set to true, and the conditions of `join_to_sort_minimum_perkey_rows` and `join_to_sort_maximum_table_rows` are met, rerange the right table by key to improve the performance in left or inner hash join"}
+            {"allow_experimental_join_right_table_sorting", false, false, "If it is set to true, and the conditions of `join_to_sort_minimum_perkey_rows` and `join_to_sort_maximum_table_rows` are met, rerange the right table by key to improve the performance in left or inner hash join"},
+            {"mongodb_throw_on_unsupported_query", false, true, "New setting."},
```
Comment (Algunenano) [inline src/Core/SettingsChangesHistory.cpp:89]: This should be in 24.10

Comment (Algunenano) [reply src/Core/SettingsChangesHistory.cpp:89]: https://github.com/ClickHouse/ClickHouse/pull/69913

**Comment by yariks5s** [comment]
I have this segfault in TableFunctionMongoDB:

```
(version 24.8.1.10338 (official build), build id: 3CA873258A46DC00BB708FDE4D68F10BF7494C07, git hash: 1463c1f269f8b41656fea6aa399bf80ec8060574) (from thread 862) (query_id: c4ebbcb7-d443-4273-a049-fa0d5cad5ac8) (query: SELECT * FROM mongodb('mongodb://movedb.cluster-cohxzo0c14le.eu-west-1.docdb.amazonaws.com:27017/?retryWrites=false&readPreference=secondaryPreferred&replicaSet=rs0&tls=true&tlsCAFile=%2Fhome%2Fkevin%2Fmongo%2Fglobal-bundle.pem', 'deck_gotest', 'ppvs', 'MoveSolutions', '[HIDDEN]', NULL)) Received signal Segmentation fault (11)

Address: 0xdf. Access: . Address not mapped to object.

0. signalHandler(int, siginfo_t*, void*)
1. ?
2. DB::(anonymous namespace)::TableFunctionMongoDB::parseArguments(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>)
3. DB::QueryAnalyzer::resolveTableFunction(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&, bool)
4. DB::QueryAnalyzer::resolveQueryJoinTreeNode(std::shared_ptr<DB::IQueryTreeNode>&, DB::IdentifierResolveScope&, DB::QueryExpressionsAliasVisitor&)
5. DB::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::IdentifierResolveScope&)
6. DB::QueryAnalyzer::resolve(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::IQueryTreeNode> const&, std::shared_ptr<DB::Context const>)
7. DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>)
8. DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>)
9. DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) (.llvm.9215533484892928605)
10. DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&)
11. std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> std::__function::__policy_invoker<std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>::__call_impl<std::__function::__default_alloc_func<DB::registerInterpreterSelectQueryAnalyzer(DB::InterpreterFactory&)::$_0, std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>>(std::__function::__policy_storage const*, DB::InterpreterFactory::Arguments const&) (.llvm.9215533484892928605)
12. DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&)
13. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
14. DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::shared_ptr<DB::Context>, std::function<void (DB::QueryResultDetails const&)>, DB::QueryFlags, std::optional<DB::FormatSettings> const&, std::function<void (DB::IOutputFormat&, String const&, std::shared_ptr<DB::Context const> const&, std::optional<DB::FormatSettings> const&)>)
15. DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::optional<DB::CurrentThread::QueryScope>&, StrongTypedef<unsigned long, ProfileEvents::EventTag> const&)
16. DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&, StrongTypedef<unsigned long, ProfileEvents::EventTag> const&)
17. DB::HTTPServerConnection::run()
18. DB::TCPProtocolStackHandler::run()
19. Poco::Net::TCPServerConnection::start()
20. Poco::Net::TCPServerDispatcher::run()
21. Poco::PooledThread::run()
22. Poco::ThreadImpl::runnableEntry(void*)
23. ?
24. ?
```

It can be connected to this PR.

**Comment by yariks5s** [comment]
cc @vdimir

**Comment by clickgapai** [comment]
Hi @vdimir — while reviewing this PR I found the following:

- **Bug (P2):** `getKeyValueMongoDBArgument` rejects `oid_columns` key-value argument despite being documented — https://github.com/ClickHouse/ClickHouse/issues/105033

Happy to discuss — close anything that's wrong or already addressed.

**Comment by clickgapai** [comment]
Hi @vdimir — a couple of suggestions to make this PR easier for reviewers and the changelog reader.

**Title** (currently `vague`):
- Current: `Resubmit #63279`
- Why: "Resubmit #63279" tells the reader nothing about what the PR does. It requires opening the referenced PR to understand that this adds a new MongoDB integration based on the official mongo-cxx-driver.
- Suggested: `Add MongoDB integration based on the official mongo-cxx-driver`

**Description** (currently `incomplete`):
- Why: The body is the CI template only; it does not describe that the PR adds new mongo-c/mongo-cxx submodules, a new StorageMongoDB/table-function implementation gated by USE_MONGODB, and a `use_legacy_mongodb_integration` server setting that retains the old Poco path by default.
- Suggested addition:

  > This resubmits #63279: it adds a MongoDB integration built on the official mongo-c-driver / mongo-cxx-driver (new submodules, built statically under USE_MONGODB). The legacy Poco-based implementation is retained and remains the default via the `use_legacy_mongodb_integration` server setting; the new implementation is opt-in. Includes BSON<->ClickHouse type conversion (BSONCXXHelper), URI credential masking, and integration tests for the storage engine, table function, and dictionary source.

Ignore this if the current title/body is intentional.

**Pull request merged**